plugins/mirror/bazel_test.go

```diff
@@ -161,9 +216,230 @@ func TestRemoveStaleDownloadURLS(t *testing.T) {
+	{
+		name: "has urls",
+		data: []byte(`
+rpm(
+    name = "findutils-1__4.7.0-4.fc32.x86_64",
+    sha256 = "c7e5d5de11d4c791596ca39d1587c50caba0e06f12a7c24c5d40421d291cd661",
+    urls = [
+        "https://mirror.dogado.de/fedora/linux/updates/32/Everything/x86_64/Packages/f/findutils-4.7.0-4.fc32.x86_64.rpm",
+    ],
+)
+`),
+		expectFails: false,
+	},
+	{
+		name: "has url",
+		data: []byte(`
+rpm(
+    name = "findutils-1__4.7.0-4.fc32.x86_64",
+    sha256 = "c7e5d5de11d4c791596ca39d1587c50caba0e06f12a7c24c5d40421d291cd661",
+    url = "https://mirror.dogado.de/fedora/linux/updates/32/Everything/x86_64/Packages/f/findutils-4.7.0-4.fc32.x86_64.rpm",
+)
+`),
+		expectFails: false,
+	},
+	{
+		name: "neither urls nor url",
+		data: []byte(`
+rpm(
+    name = "findutils-1__4.7.0-4.fc32.x86_64",
+    sha256 = "c7e5d5de11d4c791596ca39d1587c50caba0e06f12a7c24c5d40421d291cd661",
+)
+`),
+		expectFails: true,
+	},
+	{
+		name: "urls empty",
+		data: []byte(`
+rpm(
+    name = "findutils-1",
+    sha256 = "c7e5d5de11d4c791596ca39d1587c50caba0e06f12a7c24c5d40421d291cd661",
+    urls = [],
+)
+`),
+		expectFails: true,
+	},
+	{
+		name: "two artifacts with urls empty, both names should appear in error message",
+		data: []byte(`
+rpm(
+    name = "findutils-1",
+    sha256 = "c7e5d5de11d4c791596ca39d1587c50caba0e06f12a7c24c5d40421d291cd661",
+    urls = [],
+)
+rpm(
+    name = "findutils-2",
+    sha256 = "c7e5d5de11d4c791596ca39d1587c50caba0e06f12a7c24c5d40421d291cd662",
+    urls = [],
+)
+`),
+		expectFails: true,
+	},
+}
+
+func TestCheckArtifactsHaveURLS(t *testing.T) {
+	for _, workspaceData := range testCasesForCheckArtifactsHaveURLs {
+		file, err := build.ParseWorkspace("workspace", workspaceData.data)
+		if err != nil {
+			t.Fatal(err)
+		}
+		artifacts, err := GetArtifacts(file)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		err = CheckArtifactsHaveURLS(artifacts)
+		if workspaceData.expectFails {
+			if err == nil {
+				t.Fatalf("'%s': expected check to fail, URLS: %v", workspaceData.name, artifacts[0].URLs())
+			}
+			for _, shouldContain := range workspaceData.shouldContain {
+				if !strings.Contains(err.Error(), shouldContain) {
+					t.Fatalf("'%s': expected error to contain %s, was: %s", workspaceData.name, shouldContain, err.Error())
+				}
+			}
+		} else if !workspaceData.expectFails {
+			if err != nil {
+				t.Fatalf("'%s': expected check not to fail, URLS: %v", workspaceData.name, artifacts[0].URLs())
+			}
+		}
+	}
+}
+
+type artifactAppendURLTestData struct {
+	artifact     *Artifact
+	newURL 	     string
+	expectedURLs []string
+}
+
+func (t *artifactAppendURLTestData) HasExpectedURLs() bool {
+	if len(t.expectedURLs) != len(t.artifact.URLs()) {
+		return false
+	}
+	for i, v := range t.expectedURLs {
+		if v != t.artifact.URLs()[i] {
+			return false
+		}
+	}
+	return true
+
+}
+
+var artifactAppendURLTestDataSet = []artifactAppendURLTestData{
+	// No url data present
```

Comment:
It would be great to define a `description` field in `artifactAppendURLTestData` and add this strings as part of each test case. This would be more useful using subtests in `TestArtifact_AppendURL`, wdyt? It would look like:
```
func TestArtifact_AppendURL(t *testing.T) {
	for _, testData := range artifactAppendURLTestDataSet {
                t.Run(testData.description, func(t *testing.T) {
		        testData.artifact.AppendURL(testData.newURL)
		        if !testData.HasExpectedURLs() {
			        t.Fatalf("expected: %v, actual: %v", testData.expectedURLs, testData.artifact.URLs())
		        }
	        }
        }
}
```